docs/04.guides/04.cookbooks/29.Threads-Scope/page.md

```diff
@@ -0,0 +1,272 @@
+
+Threads were mainly used to retrieve the data form DB, cfhttp, webservice it doesn't care about how much time it takes to execute.
+
+### Example 2 ###
+
+Here we can see about multiple threads in lucee. Lucee can also used to create multiple threads running parallel at that time.
+
+```lucee
+<cfscript>
+function napASecond(index) localmode=true {
+	thread  {
+		thread.start=now();
+		sleep(1000)
+		thread.end=now();
+	}
+}
+
+start=getTickCount();
+loop from=1 to=5 index="index" {
+	napASecond(index);
+}
+
+// show threads
+dump(var:cfthread,expand:false);
+// wait for all threads to finish
+thread action="join" name=cfthread.keyList();
+// show threads
+dump(var:cfthread,expand:false);
+
+dump("done in #getTickCount()-start#ms");
+</cfscript>
+```
+
+Above example shows currently five threads were running parallel, Thread action join wait for all threads to finish.
+
+```cfthread``` returns struct info of all thread status.
+
+
+Same example within tag
+
+```lucee
+<cffunction name="napASecond">
+	<cfargument name="index">
+	<cfthread index="#index#" action="run" name="t#index#">
+		<cfset thread.start=now()>
+		<cfset sleep(1000+(index*10))>
+		<cfset thread.end=now()>
+	</cfthread>	
+</cffunction>
+
+<cfset start=getTickCount()>
+
+<cfloop from="1" to="5" index="index">
+	<cfset napASecond(index)>
+</cfloop>
+
+<!-- show threads -->
+<cfdump var="#cfthread#" expand="false">
+
+<!-- wait for all threads to finish -->
+<cfthread action="join" name="t1,t2,t3">
+
+<!-- show threads -->
+<cfdump var="#cfthread#" expand="false">
+
+<cfdump var="done in #getTickCount()-start#ms">
+
+```
+
+Here it join three threads "t1,t2,t3". We can join threads by its name.
+
+
+### Example 3 ###
+
+Here we can see about threads running parallel using ```each``` function.
+
+```lucee
+<cfscript>
+tasks = ["Susi","Sorglos"];
+start=getTickCount();
+tasks.each(
+	function(value, index, array){
+		systemOutput(value,true);
+		sleep(1000);
+	}
+	,true
+);
+dump("done in #getTickCount()-start#ms");
+</cfscript>
+```
+
+```each``` function has a option parallel, when we set parallel attribute as true. It runs each element parallel. In default it should be false.
+
+So that two array elements runs parallel done in 1second.
+
+
+```lucee
+<cfscript>
+tasks = [];
+loop from=1 to=100 index="i" {
+	arrayAppend(tasks,"t"&i)
+}
+start=getTickCount();
+tasks.each(
+	function(value, index, array){
+		systemOutput(value,true);
+		sleep(1000);
+	}
+	,true
+	//,100
+);
+dump("done in #getTickCount()-start#ms");
+</cfscript>
+```
+
+Here we have 100 array elements, setting option parallel as true it execute for 5 seconds not at 1 second because in default lucee runs 20 threads at the time after that it wait for free threads so only it take 5 seconds to complete.
+
+But we have option to set maximum threads open at that time. If it set to 100, it can open 100 threads at that time run parallel it take 1 second to complete.
```

Comment:
***ToChange***
***This is for lines 154 through 156***
Here we have 100 array elements. Setting the optional attribute parallel as true, it executes in 5 seconds - not in 1 second - because by default, Lucee runs 20 threads at a time and after that, it waits for free threads. So, it takes 5 seconds to complete.

But we have the option to set the maximum threads open at a time. If it set to 100, Lucee can open 100 threads at a time to run in parallel. Then this code only takes 1 second to complete.